### ISSUE TITLE
npm scripts shell select

### ISSUE THREAD

**Issue by XVincentX** [npm/npm#6543]
Good morning sirs,
I would like to understand if npm scripts under script section of _packages.json_ are run using the shell listed into configuration 

> npm config get shell

From my tests (SO: Windows), looks like that even if **sh** have been set (--global flag), cmd.exe is being used and so it is unable to perform my tasks.

I've made an overview of documentation and source code too, but I was not able to figure out.

Thank you,
V.

**Comment by majgis** [comment]
I was very surprised to find that when using cygwin, git bash, or PowerShell, `npm run` still executes all commands in cmd.exe.   This behavior seems like a serious flaw with npm.

**Comment by majgis** [comment]
I deleted this [if statement](https://github.com/npm/npm/blob/master/lib/utils/lifecycle.js#L17) and executed `npm run`.  It behaved as I expected under cygwin, git bash and powershell.

**Comment by majgis** [comment]
I can see two issues with using NPM_CONFIG_SHELL for run scripts on windows:

1. This would be a breaking change
2. The args for spawn are different for cmd vs bash, so the shell type would still need to be identified on windows.

I explored the environment variables output from `process.env` across PowerShell, git bash and cygwin and I couldn't find anything consistent to hook into.

**Proposal:**
Create a new npm config setting called `posix`.
* The command `npm config set posix true` sets posix=true in .npmrc
* The default value in windows is `false`, and it is `true` on all other platforms
* Those using cygwin, powershell, git bash or other POSIX shell can set NPM_CONFIG_POSIX=true
*  Any other behavior, in addition to run scripts, where it would make sense to flip to the POSIX default behavior should also be changed.
* Existing projects need only add the following to their package.json to make their project POSIX friendly, and make it explicit what syntax is being used:

        "config": {
             "posix":true
        }

**Comment by XVincentX** [comment]
I have no knowledge at all on nom code (so I am not the right person to evaluate your proposal), but I like your solution, actually.

**Comment by othiym23** [comment]
If – and this is a big if – a way can be found to support this that doesn't _ever_ break things for Windows users following the path of least resistance (i.e. running npm from `cmd.exe`, using the npm included by default with Node, installed via an official Node.js or io.js installer), then I could see it landing.

I have very strong motivation to be somewhat strict / conservative about this. It is hard to overstate how much of the support burden for npm comes from those users, and I'm very reluctant to risk breaking things for this particular audience. I know that there are more than a few longtime Node developers running on Windows via mingw2 or using the Cygwin-bundled bash, and I get how frustrating this limitation must be for you, but there are many many many more Windows developers who are relatively inexperienced with Node and Posixy things who just want to ensure that `ember-cli` or Cordova continue to work.

To be candid, the two things in npm that freak me out the most are tinkering with the Windows integration and messing with how `npm-registry-client` deals with the network, because whenever either of those things break, it has an immediate and tangible impact on npm's users in corporate IT environments, and I know about it **immediately**. So, I'm not saying we can't improve this experience, but I *am* saying that whatever happens needs to be very carefully implemented and thoroughly tested before it's going to be incorporated into npm.

**Comment by majgis** [comment]
I will work on creating a PR, well knowing it is still a big if.

**Option 1**
Proceed with the requirement that if `posix` is false or undefined, the current system default shell is used. When `posix` is true, the shell is changed from the system default to `sh`.  In the case of Linux  or Mac, the system default is sh, so there is no change.  In the case of Windows, the system default is cmd, so it is changed to sh.


**Option 2**
To make this config setting a little more interesting for those with POSIX as their system default, we could use `posix-shell` instead of `posix`.  If set to true, the same behavior as option 1 is used.  In addition, a name to a shell on PATH can be used, ie. ksh or bash,  or a path to a shell can be used.

Let me know which option you prefer.  I prefer option 2, since option 1 is really a windows specific setting, and who doesn't want more control.  Although option 2 does add complexity,

**Comment by XVincentX** [comment]
I'd vote for option 2.

**Comment by othiym23** [comment]
I think I'd prefer a more generic name, like `command-shell`, because there's really only one POSIX shell. Aside from that nit, option 2 is my preference as well.

**Comment by majgis** [comment]
I agree, posix is incorrect.  Using command-shell might be a little too general, as it brings to mind cmd on Windows.

How about [unix-shell](http://en.wikipedia.org/wiki/Unix_shell)?

I am proceeding with option 2.

**Comment by CaitlinWeb** [comment]
What is the status of this? 

I find it quite annoying that, even though I'm running the command originally in Git Bash or Powershell, `npm` runs the command in cmd.exe instead. This means globs don't work, even though they would if `npm` would run `scripts` commands in the shell I'm currently using.

It would be nice if there was a configuration option to tell `npm` to run `scripts` with whatever you set it to, similar to `npm config set shell "C:\Program Files\Git\git-bash.exe"`.

**Comment by majgis** [comment]
I will step down.  This is open to anyone interested.  I changed jobs and
am now using a mac.  I thought I would still pull it off on a junky old
laptop, but my personal interest has waned, so isn't going to happen.  I do
think it is an important feature for windows users and I hope someone
follows through...

**Comment by njam3** [comment]
Running into this issue now, seems like being able to specify the shell on windows while defaulting to COMSPEC would be an easy option. We are trying to create a unified build process via npm scripts and windows users unable to use bash is complicating that process. Especially since we can't set node environment variables within an npm script that works for all environments (export command doesn't work in cmd.exe and set doesn't work in OSX for this application)

**Comment by dchowitz** [comment]
Also stumbled upon this issue. My case: I have a Mac Book running an outdated bash (3.2.57) and I would like to use npm to use the actual version (installed with brew).

**Comment by othiym23** [comment]
The CLI team did a little belated spelunking here and realized that the only place the existing `shell` configuration value is used is for `npm explore`, which is probably not the npm command that most people would associate with that configuration name. Making this configurable still seems like a good idea, and there have been a few attempts at PRs to get this in, but as the discussion above attests, there’s still some confusion and subtlety involved in changing how this piece of npm works. The team would still like to see this made more configurable, and I imagine a fair number of Windows users would as well. As such, the team will get to this… eventually. That may or may not be within the next 6-12 months, so if some of you are feeling a more urgent need for this functionality, putting together a patch for us to review and merge would be very helpful. Thanks for your time and patience!

**Comment by XVincentX** [comment]
That sucks, but I understand that it might be difficult. Will continue to workaround it as long it will be necessary :(

**Comment by ForsakenHarmony** [comment]
well, rip windows users

**Comment by elclanrs** [comment]
We got in some trouble because of this issue. I assumed that if you run `npm run script` in git bash it would execute it there too, but I was wrong. Since we couldn't run our build on Windows and we didn't want to rewrite the scripts I ended up creating a bash script at the same level as the package.json. It finds a script in the package.json (with the [json](https://www.npmjs.com/package/json) module) and evals it; it is not pretty but it works for now, and all we needed to change was calls to `npm run` (including within other scripts), to calls to `bash task`:

```sh
#!/bin/bash
export PATH="$PATH:$PWD/node_modules/.bin"
for arg in $@; do eval $(cat package.json | json "scripts['$arg']"); done
```

So to run the "build" script for example we do `bash task build` or `./task build`. On unix systems you can still run `npm run build` but scripts that use other scripts should still use `bash task` insted of `npm run` because we end up with the same problem otherwise.

**Comment by fractalawareness** [comment]
Have updated Node (and npm) recently and got this issue. Windows users are sad now 😞

**Comment by SerG-Y** [comment]
Hi, guys, I faced this problem in windows 10 again. Can someone help me? 
I tried this solution https://stackoverflow.com/a/25614832 but it didn't help me and I tried to set `comspec` like this `npm config set comspec="C:\\Program Files\\Git\\bin\\bash.exe"` it didn't fix too.

**Comment by madarche** [comment]
@SerG-Y the solution is here in #16687 and [it's available in npm >= 5.1.0](https://github.com/npm/npm/releases/tag/v5.1.0)

And that's why this issue is closed.
